src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,424 @@
+    # 2×2 DataFrame
+    # │ Row │ a     │ b     │
+    # │     │ Int64 │ Int64 │
+    # ├─────┼───────┼───────┤
+    # │ 1   │ 1     │ 3     │
+    # │ 2   │ 2     │ 4     │
+    #
+    # julia> select(df, :, :a=>sin=>:a, :a, 1)
+    # 2×2 DataFrame
+    # │ Row │ a        │ b     │
+    # │     │ Float64  │ Int64 │
+    # ├─────┼──────────┼───────┤
+    # │ 1   │ 0.841471 │ 3     │
+    # │ 2   │ 0.909297 │ 4     │
+    #
+    # we compute column :a immediately when we process `:` although it is specified
+    # later by `:a=>sin=>:a` because we know from `transformed_cols` variable that
+    # it will be computed later via a transformation
+    transformed_cols = Dict{Symbol, Any}()
+    for nc in normalized_cs
+        if nc isa Pair
+            newname = last(last(nc))
+            @assert newname isa Symbol
+            if haskey(transformed_cols, newname)
+                throw(ArgumentError("duplicate target transformed or renamed " *
+                                    "column name $newname passed"))
+            end
+            transformed_cols[newname] = nc
+        end
+    end
+    for nc in normalized_cs
+        if nc isa Union{Int, AbstractVector{Int}}
+            allunique(nc) || throw(ArgumentError("duplicate column names selected"))
+            for i in nc
+                newname = _names(df)[i]
+                if !hasproperty(newdf, newname)
+                    if haskey(transformed_cols, newname)
+                        nct = transformed_cols[newname]
+                        @assert nct !== nothing
+                        select_transform!(nct, df, newdf, transformed_cols, copycols)
+                    else
+                        newdf[!, newname] = copycols ? df[:, i] : df[!, i]
+                    end
+                end
+            end
+        else
+            select_transform!(nc, df, newdf, transformed_cols, copycols)
+        end
+    end
+    return newdf
+end
+
+select(dfv::SubDataFrame, ind::ColumnIndex; copycols::Bool=true) =
+    select(dfv, [ind], copycols=copycols)
+select(dfv::SubDataFrame, inds::Union{AbstractVector{<:Integer}, AbstractVector{Symbol},
+                                      Colon, All, Not, Between, Regex}; copycols::Bool=true) =
+    copycols ? dfv[:, inds] : view(dfv, :, inds)
+
+function select(dfv::SubDataFrame, inds...; copycols::Bool=true)
+    if copycols
+        return _select(dfv, [normalize_selection(index(dfv), c) for c in inds], true)
+    else
+        # we do not support transformations here
+        # newinds should not be large so making it Vector{Any} should be OK
+        newinds = []
+        seen_single_column = Set{Int}()
+        for ind in inds
+            if ind isa ColumnIndex
+                ind_idx = index(dfv)[ind]
+                if ind_idx in seen_single_column
+                    throw(ArgumentError("passing duplicates of single column selection" *
+                                        "in `select` is not allowed"))
```

Comment:
Print `ind`?

Also I'd say "selecting the same column multiple times using symbol or integer is not allowed".

Reply:
fixed